filebeat/input/filestream/input_integration_test.go

```diff
@@ -100,3 +147,223 @@ func TestFilestreamCloseEOF(t *testing.T) {
+	})
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	testlines := []byte("first log line\nnext is an empty line\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	env.waitUntilEventCount(2)
+	env.requireOffsetInRegistry(testlogName, len(testlines))
+
+	moreTestlines := []byte("\nafter an empty line\n")
+	env.mustAppendLinesToFile(testlogName, moreTestlines)
+
+	env.waitUntilEventCount(3)
+
+	cancelInput()
+	env.waitUntilInputStops()
+
+	env.requireOffsetInRegistry(testlogName, len(testlines)+len(moreTestlines))
+}
+
+// test_empty_lines_only from test_harvester.py
+// This test differs from the original because in filestream
+// input offset is no longer persisted when the line is empty.
+func TestFilestreamEmptyLinesOnly(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths":                             []string{env.abspath(testlogName)},
+		"prospector.scanner.check_interval": "1ms",
+	})
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	testlines := []byte("\n\n\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	cancelInput()
+	env.waitUntilInputStops()
+
+	env.requireNoEntryInRegistry(testlogName)
+}
+
+// test_exceed_buffer from test_harvester.py
+func TestFilestreamExceedBuffer(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths":       []string{env.abspath(testlogName)},
+		"buffer_size": 10,
+	})
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	testline := []byte("a line longer than size allowed in buffer_size\n")
+	expectedOffset := len(testline)
+	env.mustWriteLinesToFile(testlogName, testline)
+
+	// event arrives to the output in full
+	env.waitUntilEventCount(1)
+	env.requireEventsReceived([]string{string(testline[:len(testline)-1])})
+
+	cancelInput()
+	env.waitUntilInputStops()
+
+	env.requireOffsetInRegistry(testlogName, expectedOffset)
+}
+
+// test_truncated_file_open from test_harvester.py
+func TestFilestreamTruncatedFileOpen(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths":                             []string{env.abspath(testlogName)},
+		"prospector.scanner.check_interval": "1ms",
+	})
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	testlines := []byte("first line\nsecond line\nthird line\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	env.waitUntilEventCount(3)
+	env.requireOffsetInRegistry(testlogName, len(testlines))
+
+	env.mustTruncateFile(testlogName, 0)
+
+	truncatedTestLines := []byte("truncated first line\n")
+	env.mustWriteLinesToFile(testlogName, truncatedTestLines)
+	env.waitUntilEventCount(4)
```

Comment:
This test makes me wonder, what happens if the file is bigger than before when it is finally discovered by the prospector that the file has been truncated?

This has been a common problem in the logs input (where only the harvester did detect truncation). I presume with the prospector restarting the harvvester early the chance of this happening is reduced.

Reply:
> However, if the output is blocked filestream is waiting to publish previous events. So the Harvester has to wait for the previous one to publish and stop before it can start reading from the beginning.

Hmm... One of the goals of the filestream input is to decouple control from the state of the output. Even if the output is blocked, the harvester must be reactive and shutdown if it is told to do so. As annoying as it is, but once a file is truncated, there is no way for us to recover the content of the file if lines have not yet read, or filebeat is restart. On truncate we should ensure that the pending state updates are not written to the registry anymore... they must get lost. This ensures that we can unlock the registry state immediately and start a new harvester that can start overwriting existing state immediately.

The invalidation of pending registry state updates can be implemented in the ack handler, or the update op itself (e.g. using a shared 'alive' atomic bool). For example the ack handler for the harvester is [setup here]( https://github.com/elastic/beats/blob/0b4e053f5d954b1bfa960cc1800e23d2c4ee6b28/filebeat/input/filestream/internal/input-logfile/input.go#L84). Each harvester has its own beat.Client, including its private ACK handling. An atomic bool can be introduced to tell the acker to discard updateOps. The bool would be set to true if a file is truncated, meaning that all pending state is to be discarded.

Reply:
Ok, it can be solved quite easily by running `defaultHarvesterGroup.Stop` before restart. My reasoning behind not doing so was to be able to flush the events that are already picked up by Filebeat but not yet accepted by the output.

Reply:
The ack handler is async and might even outlive the harvester or harvester group. The ack handler lives for as long as there are pending events in the queue to be ACKed.
The events that have already been enqueued will still be shippped. They are not removed. At worst only the current event to be published will be discarded, because the output as blocking publishing. But as the truncate operation is racy anyways I think that is an OK compromise.

Events for a single resource are always ACKed in order. That is when just calling stop and then starting a new harvester, we still will see the offset in the registry to increase with the old events before it is reset.

Just calling `Stop` would be good enough indeed. But as the offset is still updated, it is impossible for the prospector to reset the read offset to 0, as there might be pending events still updating the offset on disk. Unlinking the ack handler/updateOp in case of a truncate event allows us to drop pending updates, set the read offset in the persistent store to 0 and start a new harvester. Setting the read to 0 in the persisted store improves resiliency in case Filebeat is restarted. If Filebeat is not restarted, everything will be fine. Auto discovery adding/removing the filestream input for said files will have no impact as well.